File: Assets/XLuaExamples/11_Dispose/DisposeTest.lua
--[==[
- Author:       ezhex1991@outlook.com
- CreateTime:   2019-11-25 14:04:22
- Organization: #ORGANIZATION#
- Description:  
--]==]
local M = {}
----- CODE -----
local DisposeTest = CS.EZhex1991.EZUnity.XLuaExample.DisposeTest
local testObject = CS.UnityEngine.GameObject.Find("DisposeTest")
                        :GetComponent("EZhex1991.EZUnity.XLuaExample.DisposeTest")

local LuaFunction = function()
    print("Lua Function")
end

-- xlua.util.createdelegate的使用，使用一个C#方法在lua上创建一个C#Delegate
-- 五个参数分别是：delegate的类型，C#方法所作用的实例，实例的类型，方法的名称，参数的类型列表
-- 如果是静态方法，那么“C#方法所作用的实例”使用nil，如果方法无参，那么“参数的类型列表”空着即可
local CSharpDelegate = require("xlua.util").createdelegate(
    CS.System.Action,
    testObject,
    DisposeTest,
    "CSharpFunction",
    {}
)

-- 注册
function Register()
    print("----------Register----------");
    -- 先判断是否为nil才能用'+'
    -- if testObject.testAction == nil then
    --     testObject.testAction = LuaFunction
    --     testObject.testAction = testObject.testAction + CSharpDelegate
    -- else
    --     testObject.testAction = testObject.testAction + LuaFunction
    --     testObject.testAction = testObject.testAction + CSharpDelegate
    -- end

    -- testObject:testEvent("+", LuaFunction)
    -- testObject:testEvent("+", CSharpDelegate)
end

-- 注销
function Unregister()
    print("----------Unregister----------");
    -- if testObject.testAction ~= nil then 
    --     testObject.testAction = testObject.testAction - LuaFunction
    -- end
    -- if testObject.testAction ~= nil then 
    --     testObject.testAction = testObject.testAction - CSharpDelegate
    -- end

    -- testObject:testEvent("-", LuaFunction)
    -- testObject:testEvent("-", CSharpDelegate)

    testObject.button_Register.onClick:RemoveListener(Register)
    testObject.button_Unregister.onClick:RemoveListener(Unregister)
end

testObject.button_Register.onClick:AddListener(Register)
testObject.button_Unregister.onClick:AddListener(Unregister)
----- CODE -----
return M

File: Assets/XLuaExamples/11_Dispose/DisposeTest.cs
/* Author:          ezhex1991@outlook.com
 * CreateTime:      2019-11-25 13:36:11
 * Organization:    #ORGANIZATION#
 * Description:     
 */
using System;
using System.Collections.Generic;
using UnityEngine;
using XLua;

namespace EZhex1991.EZUnity.XLuaExample
{
    /// <summary>
    /// -----Dispose并不是必要的-----
    /// -----Dispose并不是必要的-----
    /// -----Dispose并不是必要的-----
    /// 通常来说，整个应用只需要一个LuaEnv，并且这个LuaEnv的生命周期与该应用一致，换句话说，Dispose大部分时候意味着应用要退出了
    /// 你根本不用在乎这个LuaEnv在退出前是否被合理释放 - 它是应用的一部分，自然会随着应用的结束而被系统处理
    /// 
    /// 如果你真的要手动Dispose，你只需要释放掉被引用的lua方法即可，具体说明请查看官方FAQ
    /// </summary>
    [LuaCallCSharp]
    public class DisposeTest : LuaManager
    {
        public static Action testAction;
        public static event Action testEvent;

        public static Action unregister;

        private void Start()
        {
            unregister = luaEnv.Global.Get<Action>("Unregister");

            unregister();
            // C#变量引用的lua function置为null
            unregister = null;
        }

        private void OnDestroy()
        {
            luaEnv.Dispose();
            print("LuaEnv Disposed");
        }

        public static void TestFunction()
        {
            print("C# Function");
        }
    }

    public static class DisposeTestConfig
    {
        public static List<Type> CSharpCallLua = new List<Type>()
        {
            typeof(Action),
        };
    }
}
